velox/functions/prestosql/GeometryUtils.h

```diff
@@ -0,0 +1,685 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <geos/geom/GeometryFactory.h>
+#include <geos/geom/LineString.h>
+#include <geos/geom/Point.h>
+
+#include <velox/common/base/Exceptions.h>
+#include <velox/type/StringView.h>
+
+namespace facebook::velox::functions {
+
+class GeometryUtils {
+ public:
+  template <typename StringWriter>
+  static void serialize(
+      const std::unique_ptr<geos::geom::Geometry>& geometry,
+      StringWriter& stringWriter) {
+    SliceOutput sliceOutput(stringWriter);
+    serialize(geometry.get(), sliceOutput);
+  }
+
+  static std::unique_ptr<geos::geom::Geometry> deserialize(
+      const StringView& geometryString) {
+    SliceInput sliceInput(geometryString.data(), geometryString.size());
+    return deserialize(sliceInput);
+  }
+
+  static std::unique_ptr<geos::geom::Geometry> createPoint(double x, double y) {
+    geos::geom::Coordinate coord(x, y);
+    return std::unique_ptr<geos::geom::Point>(
+        GEOMETRY_FACTORY->createPoint(coord));
+  }
+
+ private:
+  enum class GeometrySerializationType : uint8_t {
+    POINT = 0,
+    MULTI_POINT = 1,
+    LINE_STRING = 2,
+    MULTI_LINE_STRING = 3,
+    POLYGON = 4,
+    MULTI_POLYGON = 5,
+    GEOMETRY_COLLECTION = 6,
+    ENVELOPE = 7
+  };
+
+  enum class EsriShapeType : int {
+    POINT = 1,
+    POLYLINE = 3,
+    POLYGON = 5,
+    MULTI_POINT = 8
+  };
+
+  class SliceInput {
+   public:
+    SliceInput(const void* rawData, size_t dataSize)
+        : data_(reinterpret_cast<const uint8_t*>(rawData)),
+          size_(dataSize),
+          position_(0) {}
+
+    uint8_t readByte() {
+      return read<uint8_t>();
+    }
+
+    int16_t readShort() {
+      return read<int16_t>();
+    }
+
+    int32_t readInt() {
+      return read<int32_t>();
+    }
+
+    int32_t readLong() {
+      return read<int64_t>();
+    }
+
+    float readFloat() {
+      return read<float>();
+    }
+
+    double readDouble() {
+      return read<double>();
+    }
+
+    size_t remaining() const {
+      return size_ - position_;
+    }
+
+    void skip(size_t size) {
+      if (position_ + size > size_) {
+        VELOX_USER_FAIL("SliceInput read exceeds buffer size.");
+      }
+      position_ += size;
+    }
+
+   private:
+    template <typename T>
+    T read() {
+      if (position_ + sizeof(T) > size_) {
+        VELOX_USER_FAIL("SliceInput read exceeds buffer size.");
+      }
+      T value;
+      std::memcpy(&value, &data_[position_], sizeof(T));
```

Comment:
wont a reinterpret cast be better than doing a copy to value here?

Reply:
Not exactly. `reinterpret_cast` could sometimes run into memory alignment crashes.